[ZEPPELIN-465] Be able to run paragraph and the following ones

Remilito opened the pull request:
### What is this PR for?
Allow to run a paragraph **and** the ones following. This is useful for instance if you are generating a dataset in a paragraph and then have others subsequent paragraphs relying on this dataset.
In a nutshell this PR contains:
* UI: added a command in paragraph dropdown menu
* REST: added an boolean option to runparagraph (sync & async)
* updated rest doc
* z-engine: added relevant function to Note
* also refacto some duplicated code here and there.

### What type of PR is it?
Feature

### Todos
* [ ] - Once committers agree on the visual impact, update relevant documentation.

### What is the Jira issue?
[ZEPPELIN-465](https://issues.apache.org/jira/browse/ZEPPELIN-465)

### How should this be tested?
* UI: click on the "run and next ones" in the dropdown menu and verify subsequent paragraphs are also run, and not precedent ones.
* REST: post request to /api/notebook/[job|run]/$noteId/$paragraphId?subsequent=true

### Screenshots (if appropriate)
![runsubsequent](https://cloud.githubusercontent.com/assets/8293897/24159054/65bc7acc-0e5e-11e7-8e41-0549cb8e3861.png)
![confirmationdialog](https://cloud.githubusercontent.com/assets/8293897/24159053/65aa31c8-0e5e-11e7-894f-2448e103b504.png)
![paragraphrunningpending](https://cloud.githubusercontent.com/assets/8293897/24159051/6598cfc8-0e5e-11e7-9f0e-c3340351ca76.png)


### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes

jongyoul commented:
I think JobSever is more proper location to handle the job status and relationship. How do you think of it?

AhyoungRyu commented:
@Remilito Thanks for the useful feature. 
I think it'll be better to display only essential buttons in the paragraph. How about placing it under paragraph control dropdown menu instead? I mean in here 
<img width="295" alt="screen shot 2017-03-20 at 1 59 43 pm" src="https://cloud.githubusercontent.com/assets/10060731/24089082/81cb1b40-0d75-11e7-91f3-e9c8ebd3f11c.png">

Remilito commented:
@AhyoungRyu : updated (and now building and passing tests :sweat_smile: ).
@jongyoul : I am not sure to fully understand what you mean. If you are talking "UI" IMO the need to run or not a sequence of paragraphs depends mostly on the paragraphs' contents (i.e. generation of a dataset used by following paragraphs), so I find it more convenient to trigger this directly in the paragraph view rather than putting it in the job view. WDYT?

Leemoonsoo commented on zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java:
```diff
@@ -750,20 +762,31 @@ public Response runParagraphSynchronously(@PathParam("noteId") String noteId,
+        if (p.getListener() == null) {
+          note.initializeJobListenerForParagraph(p);
+        }
+        p.run();
```
calling `run()` from Paragraph directly should be avoided. Paragraph should always run through scheduler provided by interpreter. i.e `note.run(p.getId())`. 

But currently, there's no method that runs paragraph synchronously in class Note. This PR doesn't need to address, but later it's better implement method such as `note.runSynchronous(p.getId())` which submit job into scheduler one by one, instead of directly run Paragraph.run().

Remilito replied:
let's keep this for another PR.

Leemoonsoo commented on zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:
```diff
@@ -319,6 +319,30 @@ function ParagraphCtrl($scope, $rootScope, $route, $window, $routeParams, $locat
 
+  $scope.runSubsequentParagraphs = function(note,paragraph) {
+    if (!(paragraph.status in ['RUNNING','PENDING']) && paragraph.config.enabled){
+      BootstrapDialog.confirm({
```
I personally prefer skip confirmation dialog especially when runSubsequentParagraphs() is called from keyboard shortcut. What do you think? 

And after runSubsequentParagraphs() ran, paragraph lost focus, so need to use mouse again and to focus paragraph.

Remilito replied:
agreed, I'll remove the dialog when it is triggered from the keyboard shortcut.

AhyoungRyu commented on zeppelin-web/src/app/notebook/paragraph/paragraph-control.html:
```diff
@@ -133,6 +133,12 @@
         <li>
+          <a  ng-click="runSubsequentParagraphs(parentNote,paragraph)"
+              ng-show="paragraph.status!='RUNNING' && paragraph.status!='PENDING' && paragraph.config.enabled">
+            <span class="icon-control-end shortcut-icon" tooltip="Run this paragraph and the next ones"></span>Run with next ones
```
Not sure other users can notice "Run with next ones" means "Run with following all paragraphs". Can we use more intuitive name for this?

AhyoungRyu commented on docs/rest-api/rest-notebook.md:
```diff
@@ -894,6 +894,10 @@ Notebooks REST API supports the following operations: List, Create, Get, Delete,
       <td> 500 </td>
     </tr>
     <tr>
+      <td> subsequent parameter (optionnal) </td>
```
optional? :)

AhyoungRyu commented:
@Remilito Tested and it works as expected. 
Regarding docs, how about adding new section for this feature instead of `subsequent parameter (optional)` as a subsection under "Run a paragraph synchronously" & "Run a paragraph asynchronously" ? I saw many ppl have wanted this feature, so it'll be better to create new section for it. Below each "Run a paragraph synchronously" & "Run a paragraph asynchronously" section would be good place i think. What do you think?

Remilito commented:
@AhyoungRyu sorry for the late answer, I have updated the doc accordingly to your advice, and changed the wording of the UI command. It is now "Run with successors", but if a native English speaker has a better wording, any help is appreciated.

Leemoonsoo commented:
LGTM

zjffdu commented:
Here's one related PR #2195 which do it in `ZeppelinContext`, I hope we can merge these 2 tickets to use the same approach.

soralee commented:
Hi,@Remilito. I think now you needed a doing rebase.
Could you rebase it, please?
